File: src/lib/storage/migrations.ts
import { CASSANDRA_CLIENT } from './cassandra';

const MIGRATIONS = [
  [
    '000001',
    `CREATE TABLE IF NOT EXISTS mincepierank.maker (
    id text PRIMARY KEY,
    name text,
    logo text,
    website text
)`
  ],
  [
    '000002',
    `CREATE TABLE IF NOT EXISTS mincepierank.maker_pie_yearly
  (
      year int,
      makerid text,
      id text,
      displayname text,
      fresh boolean,
      labels set<text>,
      image_file text,
      web_link text,
      pack_count int,
      pack_price_in_pence int,
      PRIMARY KEY ((year, makerId, id))
  );`
  ],
  ['000003', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_yearly (year);`],
  ['000004', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_yearly (makerId);`],
  ['000005', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_yearly (fresh);`],
  ['000006', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_yearly (labels);`],
  [
    '000007',
    `CREATE TABLE IF NOT EXISTS mincepierank.maker_pie_ranking_yearly
  (
      year int,
      makerid text,
      pieid text,
      userid text,
      pastry int,
      filling int,
      topping int,
      looks int,
      value int,
      notes text,
      last_updated timestamp,
      PRIMARY KEY ((year, makerid, pieid), userid)
  );`
  ],
  ['000008', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (year);`],
  ['000009', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (makerid);`],
  ['000010', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (pieid);`],
  ['000011', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (userId);`],
  ['000012', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (pastry);`],
  ['000013', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (filling);`],
  ['000014', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (topping);`],
  ['000015', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (looks);`],
  ['000016', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (value);`],
  ['000017', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_ranking_yearly (last_updated);`],
  ['000018', `ALTER TABLE mincepierank.maker_pie_yearly ADD validated boolean;`],
  ['000019', `CREATE INDEX IF NOT EXISTS ON mincepierank.maker_pie_yearly (validated);`],
  [
    '000020',
    `CREATE TABLE IF NOT EXISTS mincepierank.year (
      year int PRIMARY KEY
    )`
  ],
  [
    '000021',
    `CREATE TABLE IF NOT EXISTS mincepierank.config (
      key text PRIMARY KEY,
      value text
    )`
  ],
  [
    '000022',
    `CREATE TABLE IF NOT EXISTS mincepierank.admins (
      id text PRIMARY KEY
    )`
  ],
  // these are just to make my life easier
  ['000023', `INSERT INTO mincepierank.year (year) VALUES (2022)`],
  ['000024', `INSERT INTO mincepierank.year (year) VALUES (2023)`],
  ['000025', `INSERT INTO mincepierank.config (key, value) VALUES ('readonly', 'true')`],
  [
    '000026',
    `CREATE TABLE IF NOT EXISTS mincepierank.user_pie_yearly
  (
      year int,
      id text,
      owner_userid text,
      maker text,
      location text,
      displayname text,
      fresh boolean,
      labels set<text>,
      image_file text,
      web_link text,
      pack_count int,
      pack_price_in_pence int,
      clean boolean,
      PRIMARY KEY ((year, id), owner_userid)
  );`
  ],
  [
    '000027',
    `CREATE TABLE IF NOT EXISTS mincepierank.user_pie_ranking_yearly
  (
      year int,
      pieid text,
      userid text,
      pastry int,
      filling int,
      topping int,
      looks int,
      value int,
      notes text,
      last_updated timestamp,
      PRIMARY KEY ((year, pieid), userid)
  );`
  ]
];

const SEEDS: [string, string[][]][] = [];

export async function runMigrations() {
  await CASSANDRA_CLIENT.execute(
    `CREATE KEYSPACE IF NOT EXISTS mincepierank WITH REPLICATION = {'class':'SimpleStrategy','replication_factor':1};`
  );
  await CASSANDRA_CLIENT.execute(
    `CREATE TABLE IF NOT EXISTS mincepierank.migrations (key text PRIMARY KEY)`
  );
  for (const [key, migration] of MIGRATIONS) {
    const res = await CASSANDRA_CLIENT.execute(
      `SELECT key FROM mincepierank.migrations WHERE key = ?`,
      [key],
      { prepare: true }
    );
    if (res.rows.length === 0) {
      console.log('Running migration: ', key);
      await CASSANDRA_CLIENT.execute(migration);
      await CASSANDRA_CLIENT.execute(
        `INSERT INTO mincepierank.migrations (key) VALUES (?)`,
        [key],
        {
          prepare: true
        }
      );
    }
  }
  for (const [seedCommand, seedEntries] of SEEDS) {
    for (const seedEntry of seedEntries)
      await CASSANDRA_CLIENT.execute(
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        `${seedCommand} values (${seedEntry.map((_) => '?').join(',')});`,
        seedEntry,
        { prepare: true }
      );
  }
}

File: internal/storage/migrations.go
package storage

import (
	"context"

	"github.com/jackc/pgx/v5"
	"github.com/rs/zerolog/log"
)

var migrations []string = []string{
	`CREATE TABLE IF NOT EXISTS migration_test_table (
		id SERIAL PRIMARY KEY,
		test_data varchar(255),
		created TIMESTAMP NOT NULL DEFAULT NOW()
	);`,
	`INSERT INTO migration_test_table (test_data) VALUES ('test migration data woo')`,
	`CREATE TABLE IF NOT EXISTS migration_test_table (
		id text PRIMARY KEY,
		name text,
		logo text,
		website text
	);`,
	`CREATE TABLE IF NOT EXISTS maker_pie_yearly
  (
      year int,
      makerid text,
      id text,
      displayname text,
      fresh boolean,
      labels text[],
      image_file text,
      web_link text,
      pack_count int,
      pack_price_in_pence int,
      PRIMARY KEY (year, makerId, id)
  );`,
	`CREATE TABLE IF NOT EXISTS maker_pie_ranking_yearly
  (
      year int,
      makerid text,
      pieid text,
      userid text,
      pastry int,
      filling int,
      topping int,
      looks int,
      value int,
      notes text,
      last_updated timestamp,
      PRIMARY KEY (year, makerid, pieid, userid)
  );`,
	`CREATE TABLE IF NOT EXISTS config (
	key text PRIMARY KEY,
	value text
  )`,
	`CREATE TABLE IF NOT EXISTS admins (
	id text PRIMARY KEY
  )`,
	`CREATE TABLE IF NOT EXISTS user_pie_yearly
  (
      year int,
      id text,
      owner_userid text,
      maker text,
      location text,
      displayname text,
      fresh boolean,
      labels text[],
      image_file text,
      web_link text,
      pack_count int,
      pack_price_in_pence int,
      clean boolean,
      PRIMARY KEY (year, id, owner_userid)
  );`,
	`CREATE TABLE IF NOT EXISTS user_pie_ranking_yearly
  (
      year int,
      pieid text,
      userid text,
      pastry int,
      filling int,
      topping int,
      looks int,
      value int,
      notes text,
      last_updated timestamp,
      PRIMARY KEY (year, pieid, userid)
  );`,
	`CREATE TABLE IF NOT EXISTS maker (
		id text PRIMARY KEY,
		name text,
		logo text,
		website text
	)`,
	`ALTER TABLE maker_pie_yearly ADD validated boolean;`,
	`CREATE TABLE IF NOT EXISTS categories (
		id uuid DEFAULT gen_random_uuid() PRIMARY KEY,
		slug text,
		label text
	)`,
	`ALTER TABLE maker_pie_yearly ADD oid uuid DEFAULT gen_random_uuid();`,
	`CREATE TABLE IF NOT EXISTS maker_pie_categories (
		maker_pie_oid uuid,
		category_id uuid,
		PRIMARY KEY (maker_pie_oid, category_id)
	)`,
	`WITH all_categories as (
	  SELECT DISTINCT(unnest(mpy.labels)) as category
	  FROM maker_pie_yearly mpy
	), slugged_categories as (
			SELECT REPLACE(category, ' ', '-') slug
	  FROM all_categories
	), distinct_slugs as (
	  SELECT DISTINCT slug
	  FROM slugged_categories
	)
	INSERT INTO categories
	select gen_random_uuid() as id, slug,
			INITCAP(REPLACE(slug, '-', ' ')) as label
	from distinct_slugs`,
	`with all_categories as (
	  SELECT DISTINCT(unnest(mpy.labels)) as category, mpy.oid
	  FROM maker_pie_yearly mpy
	), clean_categories as (
			SELECT REPLACE(category, ' ', '-') as slug, oid
	  FROM all_categories
	)
	INSERT INTO maker_pie_categories
	select mpy.oid as maker_pie_oid,
			c.id as category_id
	FROM maker_pie_yearly mpy
	inner join clean_categories cc on cc.oid = mpy.oid
	inner join categories c on c.slug=cc.slug`,
	`
	CREATE VIEW maker_pie_yearly_categories AS
		  SELECT
		    mpy.oid as oid,
		    jsonb_agg(jsonb_build_object('id', c.id, 'slug', c.slug, 'label', c.label)) as categories,
			ARRAY_AGG(c.slug) as category_slugs
		  from maker_pie_yearly mpy
		  inner join maker_pie_categories mpc on mpc.maker_pie_oid = mpy.oid
		  inner join categories c on mpc.category_id = c.id
		  group by mpy.oid;
	`,
}

var migrationLogTable string = `
	CREATE TABLE IF NOT EXISTS migrations (
		idx integer,
		enabled TIMESTAMP NOT NULL DEFAULT NOW(),
		PRIMARY KEY (idx)
	);
`

var insertLogSql string = `INSERT INTO migrations (idx) VALUES ($1)`

// Migrate implements Operations.
func (o *OperationWrapper) Migrate(c context.Context) error {
	_, err := o.db.Exec(c, migrationLogTable)
	if err != nil {
		log.Error().Err(err).Msg("Error with creating migration log table")
		return err
	}

	for idx, migration := range migrations {
		var sqldx int64
		err := o.db.QueryRow(c, "select idx from migrations where idx=$1", idx).Scan(&sqldx)
		if err == pgx.ErrNoRows {
			// have a migration to do!
			_, err := o.db.Exec(c, migration)
			if err != nil {
				log.Error().Err(err).Msgf("Error with migrating '%d'", idx)
				return err
			}
			_, err = o.db.Exec(c, insertLogSql, idx)
			if err != nil {
				log.Error().Err(err).Msgf("Error with inserting log '%d'", idx)
				return err
			}
		}
	}

	return nil
}
